### ISSUE TITLE
Unbounded backlog of Actor reminder notifications when app becomes unresponsive

### ISSUE THREAD

**Issue by cgillum** [dapr/dapr#4500]
## In what area(s)?

/area runtime

## What version of Dapr?

> 1.6.1

## Expected Behavior

If an actor has subscribed to a recurring reminder and becomes unresponsive for a period of time, after it becomes responsive again, Dapr should deliver a single reminder to the actor and only then resume its normal recurrence schedule.

## Actual Behavior

If an actor has subscribed to a recurring reminder and becomes unresponsive for a period of time, Dapr builds up a list of pending reminders and delivers all of them in a flood of notifications after the actor becomes responsive again.

I observed this while debugging an app on my local machine. I stopped my test app but kept the Dapr CLI running in the background. Several days later, I started my app again and it was bombarded with thousands of reminders. It took several minutes for Dapr to finish delivering them all.

## Steps to Reproduce the Problem

1. Start the Dapr CLI in the background
```
dapr run --app-id myapp --app-port 5000 --dapr-http-port 3500
```
2. Start the app in a separate window. I'm using C# to create an app that creates a timer which schedules the reminder - here's the full program:

```csharp

using Dapr.Actors;
using Dapr.Actors.Client;
using Dapr.Actors.Runtime;
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace DaprTesting;

static class Program
{
    static async Task Main(string[] args)
    {
        Console.WriteLine("Starting up...");

        WebApplicationBuilder builder = WebApplication.CreateBuilder(args);
        builder.Services.AddActors(options =>
        {
            options.Actors.RegisterActor<MyActor>();
        });

        WebApplication app = builder.Build();
        app.UseDeveloperExceptionPage();
        app.UseRouting();
        app.UseEndpoints(endpoints => endpoints.MapActorsHandlers());

        IHost host = app;
        await host.StartAsync();

        Console.WriteLine("Press [ENTER] to send an \"init\" message.");
        Console.ReadLine();

        ActorId actorId = new("1");
        IMyActor proxy = ActorProxy.Create<IMyActor>(actorId, nameof(MyActor));
        Console.WriteLine($"Calling DoSomethingAsync on {nameof(MyActor)}:{actorId}...");
        await proxy.DoSomethingAsync();
        Console.WriteLine();
        Console.WriteLine("Press [ENTER] to exit.");
        Console.ReadLine();
    }
}

public interface IMyActor : IActor
{
    Task DoSomethingAsync();
}

public class MyActor : Actor, IMyActor, IRemindable
{
    public MyActor(ActorHost host) 
        : base(host)
    { }

    async Task IMyActor.DoSomethingAsync()
    {
        // Deliver a reminder notification every 5 seconds
        await this.RegisterReminderAsync(
            "MyReminder",
            null,
            TimeSpan.FromSeconds(5),
            TimeSpan.FromSeconds(5));
    }

    Task IRemindable.ReceiveReminderAsync(string reminderName, byte[] state, TimeSpan dueTime, TimeSpan period)
    {
        // add breakpoint here
        this.Logger.LogWarning("[{timestamp}] Reminder {reminderName} was called!", DateTime.UtcNow.ToString("o"), reminderName);
        return Task.CompletedTask;
    }
}
```

Note that I'm using `Dapr.Actors.AspNetCore` v1.6.0 from nuget.org. For this particular program, you have to press [ENTER] to start the actor.

3. Attach a debugger to the app and set a breakpoint in the `ReceiveReminderAsync` callback, but keep the Dapr CLI running for a few minutes.

4. Remove the breakpoint and resume the app - you'll notice that many reminders were received at approximately the same time. The longer you waited, the more you'd see. Here is the output (just the warning messages):

```
[2022-04-08T23:04:43.5580182Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.5692538Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.5811929Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.5932853Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.6051341Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.6169587Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.6289579Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.6440359Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.6793758Z] Reminder MyReminder was called!
[2022-04-08T23:04:43.6995241Z] Reminder MyReminder was called!
```

FWIW, I noticed the following logs in the Dapr CLI as well. I'm not sure if it's related or helpful:

```
time="2022-04-08T16:04:43.9935951-07:00" level=warning msg="did not find address for actor ID 1 and actor type MyActor in reminder MyReminder" app_id=myapp instance=XXX scope=dapr.runtime.actor type=log ver=1.6.1
time="2022-04-08T16:04:43.9935951-07:00" level=info msg="placement tables updated, version: 36" app_id=myapp instance=XXX scope=dapr.runtime.actor.internal.placement type=log ver=1.6.1
```

## Release Note

RELEASE NOTE:

**Comment by halspang** [comment]
The way I read the documentation about reminders, this feels like the intended behavior to me. Even if an actor isn't active, a reminder is supposed to fire. If anyone depends on this behavior, which they may, this would be a breaking change for them. 

https://docs.dapr.io/developing-applications/building-blocks/actors/howto-actors/#actor-reminders

@cgillum - Do you think you could use timers instead? A timer won't fire after deactivation. The only thing is, the timer would need to be restarted after the actor comes back. However, it wouldn't build up this queue and it's easy enough to add this as part of the applications startup.

@msfussell - What do you think about this actor functionality? To me this feels a lot like a series of messages enqueueing on a pubsub system when no subscriber is active. Yes, you'll get a lot when you finally activate, but you may need each one.

**Comment by cgillum** [comment]
@halspang thanks for the comment. The odd thing is that the behavior seems to be different depending on whether the app is completely down or just unresponsive. If the app is down, I see the "expected" behavior of just one reminder. It's only when the app is "unresponsive" when I see the reminder pile-up. It feels like we should be doing one or the other.

In my case, I need reminders because I rely on them being persistent - i.e., if the process gets recycled, I need the reminder to still fire.

**Comment by vlardn** [comment]
In fact, the firing of missing reminder notifications occurred even after the system was completely down. We're deleting our test environment during the night (to save the cost) but all missed reminder calls notifications are fired immediately (of course one by one) after recreating the system in the morning. Currently, we use the actor reminder for polling, i.e. to invoke some internal system tasks. And that's not really expected behaviour for this kind of reminder usage :-/

I would suggest to extra option for reminder creation to allow ignoring of missed notifications:
```C#
Task<IActorReminder> RegisterReminderAsync(
            string reminderName,
            byte[] state,
            TimeSpan dueTime,
            TimeSpan period,
            TimeSpan ttl,
            bool skipMissedNotifications = false)
```

**Comment by vlardn** [comment]
Btw, the actor timer notifications behavior is the same. So I would suggest to add `skipMissedNotifications` parameter to RegisterTimerAsync method too:
```C#
Task<IActorReminder> RegisterTimerAsync(
            string reminderName,
            byte[] state,
            TimeSpan dueTime,
            TimeSpan period,
            TimeSpan ttl,
            bool skipMissedNotifications = false)

**Comment by olitomlinson** [comment]
I agree with @vlardn here, there should be an option when Registering the reminder to not _flood_ an Actor with reminders when it eventually comes back online after a period offline.

@yaron2 can this option be factored into the new Actor Reminder subsystem coming in 1.14?

**Comment by cicoyle** [comment]
> I agree with @vlardn here, there should be an option when Registering the reminder to not _flood_ an Actor with reminders when it eventually comes back online after a period offline.
> 
> @yaron2 can this option be factored into the new Actor Reminder subsystem coming in 1.14?

This won't be in for 1.14, and didn't make it in with the Scheduler as the underlying reminder subsystem either. I can keep this in mind for the follow-ups for Scheduler for this upcoming release. We need to add the orphaned job go routine in Scheduler for jobs that couldn't be streamed back to the sidecar along with that logic in the sidecar for jobs that couldn't be sent back to the app. I'll keep the _flooding_ in mind :)

**Comment by olitomlinson** [comment]
@JoshVanL 

I think this can be closed as Not Planned, and instead we can point people to use the new upcoming CLI commands to easily reset any unwanted reminders via `dapr scheduler delete` or `dapr workflow purge` - these commands can then run as a pre-startup script/task as part of a developers workflow

**Comment by vlardn** [comment]
We are using Microsoft.Dapr k8s-extension. How to use `dapr scheduler delete` or `dapr workflow purge` in this case?
